igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/OkHttpClientBuilder.java

```diff
@@ -0,0 +1,176 @@
+    okHttpClient.setHostnameVerifier((s, sslSession) -> true);
+    okHttpClient.setSslSocketFactory(getSslContext().getSocketFactory());
+    okHttpClient.setConnectTimeout(15, TimeUnit.SECONDS);
+    okHttpClient.setReadTimeout(15, TimeUnit.SECONDS);
+    return okHttpClient;
+  }
+
+  private static SSLContext getSslContext() {
+    SSLContext sslContext;
+    try {
+      sslContext = SSLContext.getInstance("SSL");
+      sslContext.init(null, trustAllCerts, new SecureRandom());
+    } catch (KeyManagementException | NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+    return sslContext;
+  }
+
+  public static okhttp3.OkHttpClient retryingClient3(Supplier<Token> refreshToken) {
+    return new okhttp3.OkHttpClient.Builder()
+      .addInterceptor(chain -> OkHttpClientBuilder.createRetryInterceptor3(chain, refreshToken))
+      .connectTimeout(Duration.ofSeconds(15))
+      .readTimeout(Duration.ofSeconds(15))
+      .build();
+  }
+
+  private static Response createRetryInterceptor(Interceptor.Chain chain, Supplier<Token> refreshToken) {
+    Retry retry = Retry.of("concourse.api.call", RetryConfig.custom()
+      .retryExceptions(RetryableApiException.class)
+      .build());
+
+    AtomicReference<Response> lastResponse = new AtomicReference<>();
+    try {
+      return retry.executeCallable(() -> {
+        Response response = chain.proceed(chain.request());
+        lastResponse.set(response);
+
+        switch (response.code()) {
+          case 401:
+            String body = null;
+            if(response.body() != null) {
+              BufferedSource source = response.body().source();
+              source.request(Long.MAX_VALUE); // request the entire body
+              Buffer buffer = source.buffer();
+              body = buffer.clone().readString(Charset.forName("UTF-8"));
+            }
+            if (body == null || !body.contains("Bad credentials")) {
+              response = chain.proceed(chain.request().newBuilder()
+                .header("Authorization", "bearer " + refreshToken.get().getAccessToken())
+                .build());
+              lastResponse.set(response);
+            }
+            break;
+          case 502:
+          case 503:
+          case 504:
+            // after retries fail, the response body for these status codes will get wrapped up into a ConcourseApiException
+            throw new RetryableApiException();
+        }
+
+        return response;
+      });
+    } catch(IOException e) {
+      throw new UncheckedIOException(e);
+    } catch(Exception e) {
+      return lastResponse.get();
+    }
+  }
+
+  private static okhttp3.Response createRetryInterceptor3(okhttp3.Interceptor.Chain chain, Supplier<Token> refreshToken) {
+    Retry retry = Retry.of("concourse.api.call", RetryConfig.custom()
+      .retryExceptions(RetryableApiException.class)
+      .build());
+
+    AtomicReference<okhttp3.Response> lastResponse = new AtomicReference<>();
+    try {
```

Comment:
The code inside here looks exactly like what is in createRetryInterceptor above, only wrapping it in okhttp3.Response is different.

Reply:
yes, unfortunately because of the differing types there's no way i can further generalize this.